lib/libc/common/source/time/clock.c

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2023 Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <time.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys_clock.h>
+
+#ifndef CLOCKS_PER_SEC
+#warning libc does not define CLOCKS_PER_SEC which is required by ISO
+#define CLOCKS_PER_SEC CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#endif
+
+clock_t clock(void)
+{
+	clock_t clk;
+
+	clk = ((clock_t)k_uptime_get() * CLOCKS_PER_SEC) / MSEC_PER_SEC;
```

Comment:
Couldn't the intermediate value here overflow?  If `clock_t` is signed 32-bits and CLOCKS_PER_SEC is a typical value of 1e6, then this overflows after just 2 seconds.

Reply:
That was alluded to by Keith already.

https://github.com/zephyrproject-rtos/zephyr/pull/57800#pullrequestreview-1423749926

In PicoLibc, likely, newlub, and the minimal libc (potentially others?) `clock_t` is 64-bit

Reply:
I think this is equivalent without risk of intermediate values overflowing:
```
int64_t uptime = k_uptime_get();
return uptime * (CLOCKS_PER_SEC / MSEC_PER_SEC) + (uptime * (CLOCKS_PER_SEC % MSEC_PER_SEC)) / MSEC_PER_SEC;
```
If `CLOCKS_PER_SEC % MSEC_PER_SEC == 0`, then this should reduce to a simple constant multiply, which will be faster than doing the divide at runtime.

Reply:
I was going to say that if the cast of `k_uptime_get()` was removed, then standard integer promotion should always produce a 64-bit intermediate, unsigned if `clock_t` is unsigned 64-bit and signed otherwise, which would remove the requirement that `clock_t` be 64 bit to avoid early overflow.

Of course removing the integer division is even better.  It is a shame that compilers can not make that optimization to this common pattern because they must preserve the overflow of the intermediate value, which we don't even want, much less need preserved exactly as written.

The code to handle this efficiently already exists in `z_tmcvt()`.  This could be written as:
```c
z_tmcvt(a, MSEC_PER_SEC, CLOCKS_PER_SEC, true, false, false, false);
```

This will optimize to a single multiply when possible, but can also handle cases where `CLOCKS_PER_SEC % MSEC_PER_SEC != 0`.

If `clock_t` was ever 32-bit, then I think some unnecessary 64-bit math can be avoided with:
```c
z_tmcvt(a, MSEC_PER_SEC, CLOCKS_PER_SEC, true, sizeof(clock_t) <= 4, false, false);
```

Reply:
Fixed!